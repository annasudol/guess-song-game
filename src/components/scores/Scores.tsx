import { getUserInfo, getToken } from '../../state/user/selectors';
import { useSelector } from 'react-redux';
import React, { FunctionComponent, ReactElement, useEffect, useState } from 'react';
import { useHttpClient } from '../../hooks';
import { ScoresItems } from '../../components';



import CircularProgress from '@material-ui/core/CircularProgress';

export interface ScoresDBProps {
  date: string;
  _id: string;
  points: number
  playlistId: string;
  creator: string;

}

export interface ScoresDBPropsGrouped {
  [playlistId: string]: {
    date: string;
    points: number;
  }
}

export const Scores: FunctionComponent = (): ReactElement => {
  const user = useSelector(getUserInfo);
  const [scores, setScores] = useState<ScoresDBProps[]>([]);
  const { sendRequest, isLoading } = useHttpClient();

  useEffect((): void => {
    const fetchScores = async (): Promise<void> => {
      try {
        const responseData = await sendRequest(`${process.env.REACT_APP_BACKEND_URL}/score/${user.userId}`);
        setScores(responseData.scores);
      } catch (err) {
        console.warn("cannot find user's scores", err.message);
      }
    };
    fetchScores();

  }, [sendRequest]);

  const scoresGrouped = scores?.reduce((acc: ScoresDBPropsGrouped[], item: ScoresDBProps) => {
    return [...acc, {
      [item.playlistId]: {
        points: item.points,
        date: item.date,
      }
    }]
  }, []);


  const totalScores = scoresGrouped?.reduce((acc: number, item: ScoresDBPropsGrouped) => {
    return acc + Object.values(item)[0].points;
  }, 0);

  const playlistIds = scoresGrouped.reduce((acc: string[], item: ScoresDBPropsGrouped) => {
    return [...acc, ...Object.keys(item)]
  }, [])


  if (isLoading) {
    return (
      <div className="container flex justify-center items-center full-width h-full mt-20">
        <CircularProgress size={100} color="secondary" />
      </div>
    );
  }

  return (
    <div className="container pt-10">
      <div className="flex justify-between align-top full-width">
        <h2 className="font-albaSuper text-jellyBean self-top text-2xl">hello {user.name}!</h2>
        <div className="flex justify-end w-2/5">
          <div className="text-right mr-4">
            <h2 className="font-alba text-jellyBean text-3xl -mb-2">{playlistIds?.length ? playlistIds?.length : 0}</h2>
            <p className="font-OpenSans text-light-gray text-xs opacity-50">Total Games</p>
          </div>
          <div className="text-right">
            <h2 className="font-alba text-jellyBean text-3xl -mb-2">{totalScores}</h2>
            <p className="font-OpenSans text-light-gray text-xs opacity-50">Total Points</p>
          </div>
        </div>
      </div>
      <ScoresItems playlistIds={playlistIds} scoresDB={scoresGrouped} />
    </div>
  );
};


