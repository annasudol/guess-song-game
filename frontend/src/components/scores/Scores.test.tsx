import { shallow } from 'enzyme';
import React from 'react';

import { Scores } from './Scores';
import toJson from 'enzyme-to-json';
jest.mock('react-redux');
const useSelector = require('react-redux').useSelector as jest.Mock;

describe('component', (): void => {
  describe('Scores', (): void => {
    useSelector.mockReturnValue({
      getToken: 'BQAY_Ky0EiDo4z-U-oC2rYg5aF_kcSUhbukZ7BOJkXYQJzI-ChNz0hrULaIAoTmZ6s3_W_8n2LBh6_OKm5vgtvthrWLaPnGwYutjlGFW4cwQKqvN3ovaJmksGuzj5inFTH3UpEJ6_biTyDr_0w52EBirZQLGDN42mtS7ygfv2Nz',
    });
    it('match snapshot', (): void => {
      const wrapper = shallow(<Scores />);

      expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('renders reducer info', (): void => {
      useSelector.mockReturnValue({
        name: 'test',
        image: 'simple image',
        userId: '123id',
      });
      const wrapper = shallow(<Scores />);

      expect(wrapper.find('h2.font-albaSuper').text()).toBe('hello test!');
      expect(
        wrapper
          .find('p')
          .at(0)
          .text()
      ).toBe('Total Games');
      expect(
        wrapper
          .find('p')
          .at(1)
          .text()
      ).toBe('Total Points');
    });
  });
});
