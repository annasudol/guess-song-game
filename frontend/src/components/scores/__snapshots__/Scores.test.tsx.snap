// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component Scores match snapshot 1`] = `
<div
  className="container pt-10"
>
  <div
    className="flex justify-between align-top full-width"
  >
    <h2
      className="font-albaSuper text-jellyBean self-top text-2xl"
    >
      hello 
      !
    </h2>
    <div
      className="flex justify-end w-2/5"
    >
      <div
        className="text-right mr-4"
      >
        <h3
          className="font-alba text-jellyBean text-3xl -mb-2"
        >
          0
        </h3>
        <p
          className="font-OpenSans text-light-gray text-xs opacity-50"
        >
          Total Games
        </p>
      </div>
      <div
        className="text-right"
      >
        <h3
          className="font-alba text-jellyBean text-3xl -mb-2"
        >
          0
        </h3>
        <p
          className="font-OpenSans text-light-gray text-xs opacity-50"
        >
          Total Points
        </p>
      </div>
    </div>
  </div>
  <ScoresItems
    playlistIds={Array []}
    scoresDB={Array []}
  />
</div>
`;
